### PR TITLE
[repo] Move OpenTelemetry.SemanticConventions project from main repo

### LINKED ISSUE #1657: Feature Request: Reintroduce `OpenTelemetry.SemanticConventions`
**What are you trying to achieve?**

https://github.com/open-telemetry/opentelemetry-dotnet/pull/5539 removed the package `OpenTelemetry.SemanticConventions`. This package served as a centralized resource for all standard attributes from the protocol, preventing repetition and duplication across projects. Is it possible to reintroduce this package in this repo rather than in the main one?

**Additional context.**

I'm ready to contribute to migrate the sources in this repo and update them (with the latest version of the spec).

### PR DESCRIPTION
**Pull request by joegoldman2**
Fixes #1657.

## Changes

Reintroduce `OpenTelemetry.SemanticConventions` project that has been removed with https://github.com/open-telemetry/opentelemetry-dotnet/pull/5539.
PR created as draft just to see if I'm going in the right direction ~and because the generation script is throwing an exception at the moment.~

For significant contributions please make sure you have completed the following items:

* [ ] Appropriate `CHANGELOG.md` updated for non-trivial changes
* [ ] Design discussion issue #
* [ ] Changes in public API reviewed

### CODE REVIEW

**Comment by joegoldman2** [comment]
I managed to use the generator using https://github.com/open-telemetry/opentelemetry-specification/tree/v1.13.0 but if I switch to the latest version https://github.com/open-telemetry/semantic-conventions/tree/v1.25.0, I'm getting the following exception:

```sh
$ bash generate.sh
Initialized empty Git repository in /opentelemetry-dotnet-contrib/src/OpenTelemetry.SemanticConventions/scripts/semantic-conventions/.git/
remote: Enumerating objects: 9130, done.
remote: Counting objects: 100% (3153/3153), done.
remote: Compressing objects: 100% (857/857), done.
remote: Total 9130 (delta 2629), reused 2402 (delta 2290), pack-reused 5977
Receiving objects: 100% (9130/9130), 3.18 MiB | 3.10 MiB/s, done.
Resolving deltas: 100% (6176/6176), done.
From https://github.com/open-telemetry/semantic-conventions
 * tag               v1.25.0    -> FETCH_HEAD
Updating files: 100% (326/326), done.
HEAD is now at f1be151 [chore] Prepare v1.25.0 release (#878)
Error parsing /source/general.yaml

Value 'opt_in' for required field is not allowed - @11:9
Error parsing /source/media.yaml

Invalid keys: ['stability'] - @18:15
Error parsing /source/mobile-events.yaml

Invalid keys: ['stability'] - @24:15
Traceback (most recent call last):
  File "/usr/local/bin/gen-semconv", line 8, in <module>
    sys.exit(main())
  File "/usr/local/lib/python3.8/site-packages/opentelemetry/semconv/main.py", line 67, in main
    semconv = parse_semconv(args, parser)
  File "/usr/local/lib/python3.8/site-packages/opentelemetry/semconv/main.py", line 38, in parse_semconv
    semconv.finish()
  File "/usr/local/lib/python3.8/site-packages/opentelemetry/semconv/model/semantic_convention.py", line 356, in finish
    fixpoint_ref = self.resolve_ref(semconv)
  File "/usr/local/lib/python3.8/site-packages/opentelemetry/semconv/model/semantic_convention.py", line 492, in resolve_ref
    raise ValidationError.from_yaml_pos(
opentelemetry.semconv.model.exceptions.ValidationError: Semantic Convention log-exception reference `exception.type` but it cannot be found! - @2:5
```

**Comment by joegoldman2** [comment]
Since the specs were migrated to [open-telemetry/semantic-conventions](https://github.com/open-telemetry/semantic-conventions), I haven't been able to generate a file for each signal (trace, metric, etc). I'm open to suggestions.

File: opentelemetry-dotnet-contrib.proj
Comment (CodeBlanch) [inline opentelemetry-dotnet-contrib.proj:9]: nit: Remove this. If you want to exclude `OpenTelemetry.SemanticConventions` from being packaged just add `<IsPackable>false</IsPackable>` in `OpenTelemetry.SemanticConventions.csproj`.

File: src/OpenTelemetry.SemanticConventions/scripts/generate.ps1
Comment (CodeBlanch) [inline src/OpenTelemetry.SemanticConventions/scripts/generate.ps1:4]: You have "latest" now for $GENERATOR_VERSION so this comment is lying 🤣 I think either go back to a fixed version for generator or just update the comment.

File: test/OpenTelemetry.AotCompatibility.TestApp/OpenTelemetry.AotCompatibility.TestApp.csproj
Comment (CodeBlanch) [inline test/OpenTelemetry.AotCompatibility.TestApp/OpenTelemetry.AotCompatibility.TestApp.csproj:33]: Do you really want to trim `OpenTelemetry.SemanticConventions`? Fine to do but see the note on line 13, you also have to update `ci.yml` to make sure it is fired when the source changes.

File: src/OpenTelemetry.SemanticConventions/README.md
Comment (CodeBlanch) [inline src/OpenTelemetry.SemanticConventions/README.md:9]:
```suggestion
dotnet add package OpenTelemetry.SemanticConventions --prerelease
```

File: src/OpenTelemetry.SemanticConventions/README.md
Comment (CodeBlanch) [inline src/OpenTelemetry.SemanticConventions/README.md:4]: nit: Add link to the semantic convention repo.

File: src/OpenTelemetry.SemanticConventions/PublicAPI.Shipped.txt
Comment (CodeBlanch) [inline src/OpenTelemetry.SemanticConventions/PublicAPI.Shipped.txt:1]: `PublicAPI.Shipped.txt` and `PublicAPI.Unshipped.txt` should be under a folder named `.publicApi`.

File: src/OpenTelemetry.SemanticConventions/SemanticConventionsAttributes.cs
Comment (Kielek) [inline src/OpenTelemetry.SemanticConventions/SemanticConventionsAttributes.cs:8]: Some time ago I have seen discussion to include semantic convention version in namespace to avoid breaking changes. Newer versions should just incrementally add new files.

File: src/OpenTelemetry.SemanticConventions/PublicAPI.Shipped.txt
Comment (joegoldman2) [reply src/OpenTelemetry.SemanticConventions/PublicAPI.Shipped.txt:1]: I just used the code fixer to generate these files. When moving them in `.publicApi`, the warning
`Symbol '....' is not part of the declared public API[RS0016](https://github.com/dotnet/roslyn-analyzers/blob/main/src/PublicApiAnalyzers/PublicApiAnalyzers.Help.md)` appears. Not sure what to do to fix this issue

**Review by Kielek — COMMENTED** [review]
Consider adding CI/infrqastructure code similar to https://github.com/open-telemetry/opentelemetry-dotnet-contrib/pull/1673/commits/9fbfdf2353b354a9ad0f199a14ec24da650e065a

Also it will be great to have component owner under .github repository for this project.

**Comment by joegoldman2** [comment]
> @joegoldman2
> 
> Would you mind adding a `CHANGELOG.md` file into the project? Let's seed it with something like...
> 
> ```
> # Changelog
> 
> ## Unreleased
> 
> * Updated to `1.25.0` release of OpenTelemetry Semantic Conventions.
>   ([#1672](https://github.com/open-telemetry/opentelemetry-dotnet-contrib/pull/1672))
> 
> * Moved from `https://github.com/open-telemetry/opentelemetry-dotnet/` to
>   `https://github.com/open-telemetry/opentelemetry-dotnet-contrib/`.
> ```

Done!

**Comment by joegoldman2** [comment]
> Consider adding CI/infrqastructure code similar to [9fbfdf2](https://github.com/open-telemetry/opentelemetry-dotnet-contrib/commit/9fbfdf2353b354a9ad0f199a14ec24da650e065a)
> 
> Also it will be great to have component owner under .github repository for this project.

Ok for sure, I'll do it based on this example. Thank you for your review.

File: src/OpenTelemetry.SemanticConventions/PublicAPI.Shipped.txt
Comment (joegoldman2) [reply src/OpenTelemetry.SemanticConventions/PublicAPI.Shipped.txt:1]: Done.

File: src/OpenTelemetry.SemanticConventions/SemanticConventionsAttributes.cs
Comment (CodeBlanch) [reply src/OpenTelemetry.SemanticConventions/SemanticConventionsAttributes.cs:8]: IMO a lot of work needs to be done before this is really useful 🤣 But I would suggest we separate this effort to move it here to contrib (with minor updates) and improving it further (can be done in follow-up PRs)?

Comment (Kielek) [reply src/OpenTelemetry.SemanticConventions/SemanticConventionsAttributes.cs:8]: Sure, but I would wait with re-relasing package until we have something at least usable.

Comment (CodeBlanch) [reply src/OpenTelemetry.SemanticConventions/SemanticConventionsAttributes.cs:8]: Agree with you on that 👍

**Review by lmolkova — COMMENTED** [review]
Having all attributes in one place would be problematic in the long term: different stability, huge file, hard to use

I suggest to reuse new approach we've implemented in Java and Python - group attributes by the root namespace following the discussion in https://github.com/open-telemetry/semantic-conventions/issues/551

It's being documented here https://github.com/open-telemetry/semantic-conventions/blob/18325605937cb0a15fde89faabf48c857eab0ca5/supplementary-guidelines/semantic_conventions_code_generation.md.

Here's how Java does it https://github.com/open-telemetry/semantic-conventions-java.

This is not a requirement, and from semconv perspective we want to give language SIGs a freedom to decide how they want to do it. 

Happy to help with Jinja if needed.

File: src/OpenTelemetry.SemanticConventions/scripts/templates/SemanticConventionsAttributes.cs.j2
Comment (lmolkova) [inline src/OpenTelemetry.SemanticConventions/scripts/templates/SemanticConventionsAttributes.cs.j2:32]: Once .NET allows to set schema URL, having a single one would be problematic - different instrumentations can produce different schema versions.

File: src/OpenTelemetry.SemanticConventions/scripts/templates/SemanticConventionsAttributes.cs.j2
Comment (lmolkova) [inline src/OpenTelemetry.SemanticConventions/scripts/templates/SemanticConventionsAttributes.cs.j2:37]: you might want different template attributes (like `http.request.header.<key>`) to appear differently than regular ones

File: src/OpenTelemetry.SemanticConventions/SemanticConventionsAttributes.cs
Comment (lmolkova) [inline src/OpenTelemetry.SemanticConventions/SemanticConventionsAttributes.cs:2626]: is prefixes ever necessary?

File: src/OpenTelemetry.SemanticConventions/SemanticConventionsAttributes.cs
Comment (lmolkova) [inline src/OpenTelemetry.SemanticConventions/SemanticConventionsAttributes.cs:3113]: Events will have more things to be generated, we should not keep them in semantic attributes file

File: src/OpenTelemetry.SemanticConventions/SemanticConventionsAttributes.cs
Comment (lmolkova) [inline src/OpenTelemetry.SemanticConventions/SemanticConventionsAttributes.cs:16]: `OpenTelemetry.SemanticConvetions.SemanticConventionsAttributes` seems excessive. If we decide to keep all attributes in one place, I'd suggest `OpenTelemetry.SemanticConvetions.Attributes` (`OpenTelemetry.SemanticConvetions.Metrics`, `OpenTelemetry.SemanticConvetions.Events`, etc...)

File: src/OpenTelemetry.SemanticConventions/SemanticConventionsAttributes.cs
Comment (lmolkova) [inline src/OpenTelemetry.SemanticConventions/SemanticConventionsAttributes.cs:2688]: it could make sense to exclude certain namespaces that are not relevant to .NET (like JVM metrics). Here's an example of how it's done in Java (using new per-namespace codegen) - https://github.com/open-telemetry/semantic-conventions-java/pull/54

**Comment by joegoldman2** [comment]
@Kielek should I create an empty test project for the ci to be green? https://github.com/open-telemetry/opentelemetry-dotnet-contrib/actions/runs/8812897898/job/24189525031?pr=1672

**Comment by joegoldman2** [comment]
> I suggest to reuse new approach we've implemented in Java and Python - group attributes by the root namespace following the discussion in [open-telemetry/semantic-conventions#551](https://github.com/open-telemetry/semantic-conventions/issues/551)

I will try to do it.

File: .github/workflows/package-SemanticConventions.yml
Comment (Kielek) [inline .github/workflows/package-SemanticConventions.yml:1]: Add reference to this file in `.sln`.

File: build/Projects/OpenTelemetry.SemanticConventions.proj
Comment (Kielek) [inline build/Projects/OpenTelemetry.SemanticConventions.proj:1]: Add reference to this file in `.sln`.

**Comment by Kielek** [comment]
> @Kielek should I create an empty test project for the ci to be green? https://github.com/open-telemetry/opentelemetry-dotnet-contrib/actions/runs/8812897898/job/24189525031?pr=1672

I would say that we need the possibility to disable tests for this package.
Inputs for https://github.com/open-telemetry/opentelemetry-dotnet-contrib/blob/e857280c3d5749163b94ddbdcb312bdd8438cb43/.github/workflows/Component.BuildTest.yml#L5-L23 and https://github.com/open-telemetry/opentelemetry-dotnet-contrib/blob/e857280c3d5749163b94ddbdcb312bdd8438cb43/.github/workflows/Component.Package.yml#L6-L12 should be modified with corresponding usage for semantic convention.

@CodeBlanch, do you have better advice?

File: src/OpenTelemetry.SemanticConventions/SemanticConventionsAttributes.cs
Comment (joegoldman2) [reply src/OpenTelemetry.SemanticConventions/SemanticConventionsAttributes.cs]: I removed them for the moment.

**Comment by joegoldman2** [comment]
@lmolkova I followed your recommendation and group the attributes by the root namespace as it is done for Java and Python: https://github.com/open-telemetry/opentelemetry-dotnet-contrib/pull/1672/commits/69d1326f445619fc43452e7fdd04f1f8b3d945b5. Please, let me know if it looks better.

File: src/OpenTelemetry.SemanticConventions/scripts/templates/SemanticConventionsAttributes.cs.j2
Comment (Kielek) [inline src/OpenTelemetry.SemanticConventions/scripts/templates/SemanticConventionsAttributes.cs.j2:28]: Consider [file scoped namespaces](https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/proposals/csharp-10.0/file-scoped-namespaces).

File: src/OpenTelemetry.SemanticConventions/AspnetcoreAttributes.cs
Comment (joegoldman2) [inline src/OpenTelemetry.SemanticConventions/AspnetcoreAttributes.cs:15]: I think it's worth adding a special case to rename this class to `AspNetCoreAttributes`. Same for the constants.

File: src/OpenTelemetry.SemanticConventions/scripts/generate.ps1
Comment (CodeBlanch) [inline src/OpenTelemetry.SemanticConventions/scripts/generate.ps1:23]: Since we are now generating many files, I think it would be better to drop them into a subfolder:

`-v ${ROOT_DIR}/Attributes/:/output`

Easier that way to blow them away and re-generate everything, if needed. Also it creates a nice separation between real files and auto-generated files.

File: .github/workflows/ci.yml
Comment (CodeBlanch) [inline .github/workflows/ci.yml:270]: @joegoldman2 Add this new `code-cov-name` definition (as `unittests-SemanticConventions`) in here: https://github.com/open-telemetry/opentelemetry-dotnet-contrib/blob/dc7be37862e4669638bb9df42db83794cbb667d7/.github/codecov.yml#L34

**Comment by CodeBlanch** [comment]
@joegoldman2 This PR is getting pretty big. I suggest don't make any more changes to the template or try to resolve all of @lmolkova's feedback. Do those things as smaller follow-up PRs once this lands. Easier for everyone to review that way and we can document/discuss the changes in isolation.

File: src/OpenTelemetry.SemanticConventions/scripts/generate.ps1
Comment (joegoldman2) [reply src/OpenTelemetry.SemanticConventions/scripts/generate.ps1]: Done!

**Comment by joegoldman2** [comment]
> @joegoldman2 This PR is getting pretty big. I suggest don't make any more changes to the template or try to resolve all of @lmolkova's feedback. Do those things as smaller follow-up PRs once this lands. Easier for everyone to review that way and we can document/discuss the changes in isolation.

I agree, probably the rest can be done in  follow-up PRs.

File: .github/workflows/ci.yml
Comment (CodeBlanch) [reply .github/workflows/ci.yml:270]: I ended up pushing some changes to the branch to fix CI so I added this.

**Comment by CodeBlanch** [comment]
@Kielek I pushed changes here to tweak `Component.BuildTest.yml` so that it skips code coverage upload if there isn't any data available.

Also there was a typo in `codecov.yml` which is actually breaking all code coverage upload currently 😬

**Review by CodeBlanch — APPROVED** [review]
LGTM

**Review by Kielek — APPROVED** [review]

File: src/OpenTelemetry.SemanticConventions/scripts/templates/SemanticConventionsAttributes.cs.j2
Comment (lmolkova) [inline src/OpenTelemetry.SemanticConventions/scripts/templates/SemanticConventionsAttributes.cs.j2:64]: nice trick! 
I believe we gave up on it in Java and Python and do auto-formatting as a part of generate script :)

File: src/OpenTelemetry.SemanticConventions/scripts/templates/SemanticConventionsAttributes.cs.j2
Comment (lmolkova) [inline src/OpenTelemetry.SemanticConventions/scripts/templates/SemanticConventionsAttributes.cs.j2:63]: nit: it should be possible to get rid of `print_value` macro and use `print_member_value` filter provided by the otel/semconvgen

```suggestion
        public const {{ type }} {{ member.member_id | to_camelcase(True) }} = {{ print_member_value(member) }};
```

**Review by lmolkova — APPROVED** [review]
Thank you!

File: src/OpenTelemetry.SemanticConventions/scripts/templates/SemanticConventionsAttributes.cs.j2
Comment (joegoldman2) [reply src/OpenTelemetry.SemanticConventions/scripts/templates/SemanticConventionsAttributes.cs.j2]: Excellent, done!

**Comment by joaopgrassi** [comment]
Hi all, and @joegoldman2 

I'm following up on this as a semconv maintainer and helping migrate to the new tooling (Weaver). I notice that 
the script says semconv 1.25, but I see attributes that are waaaaay old, like the ones in AndroidAttributes and I'm a bit confused. 

Was this PR really updated with semconv version 1.25? Or was it just migrated as is from the main repo?

**Comment by joegoldman2** [comment]
@joaopgrassi if I remember well, I used the generate script that is part of this PR to generate the files. So normally the version 1.25 has been used.

**Comment by ben-wilson-mews** [comment]
@joegoldman2 I notice that this package isn't published to nuget, any timeline on it being published? The instructions I followed in the readme for the package.

**Pull request merged**
